sympy/core/function.py

```diff
@@ -2787,4 +2547,407 @@ def nfloat(expr, n=15, exponent=False):
         lambda x: isinstance(x, Function)))
 
 
+class AppliedFunction(Expr):
+    """
+
+    """
+
+    is_Function = True
+
+    #@cacheit
+    def __new__(cls, func, *args, **options):
+        n = len(args)
+        if n not in func.nargs:
+            # XXX: exception message must be in exactly this format to
+            # make it work with NumPy's functions like vectorize(). See,
+            # for example, https://github.com/numpy/numpy/issues/1697.
+            # The ideal solution would be just to attach metadata to
+            # the exception and change NumPy to take advantage of this.
+            temp = ('%(name)s takes %(qual)s %(args)s '
+                   'argument%(plural)s (%(given)s given)')
+            raise TypeError(temp % {
+                'name': func,
+                'qual': 'exactly' if len(func.nargs) == 1 else 'at least',
+                'args': min(func.nargs),
+                'plural': 's'*(min(func.nargs) != 1),
+                'given': n})
+
+        evaluate = options.get('evaluate', global_evaluate[0])
+
+        result = cls._application(func, args, options)
+        if evaluate and isinstance(result, AppliedFunction) and result.args:
+            pr2 = min(cls._should_evalf(func, a) for a in result.args)
+            if pr2 > 0:
+                pr = max(cls._should_evalf(func, a) for a in result.args)
+                result = result.evalf(mlib.libmpf.prec_to_dps(pr))
+
+        return result
+
+    def __getattr__(self, attr):
+        # Remapping the `_eval_*` to `self.func`:
+        if (not attr.startswith("_eval_")):
```

Comment:
I don't know if this always is correct. It works for the assumptions, but for other things I think it could be an issue.